src/Raven.Server/Documents/Handlers/BatchHandler.cs

```diff
@@ -238,45 +241,57 @@ private async Task HandleClusterTransaction(DocumentsOperationContext context, M
-            var result = await ServerStore.SendToLeaderAsync(clusterTransactionCommand);
+            var array = new DynamicJsonArray();
+            long index;
 
-            if (result.Result is List<ClusterTransactionCommand.ClusterTransactionErrorInfo> errors)
+            if (ServerStore.CommandAlreadyInLog(clusterTransactionCommand, out index, out var result, out _)) // failover
             {
-                HttpContext.Response.StatusCode = (int)HttpStatusCode.Conflict;
-                throw new ClusterTransactionConcurrencyException($"Failed to execute cluster transaction due to the following issues: {string.Join(Environment.NewLine, errors.Select(e => e.Message))}")
+                if (result == null)
                 {
-                    ConcurrencyViolations = errors.Select(e => e.Violation).ToArray()
-                };
-            }
-
-            // wait for the command to be applied on this node
-            await ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, result.Index);
+                    await ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, index); // wait for the command to be applied on this node
+                    if (Database.LastCompletedClusterTransactionIndex < index)
+                        while (await Database.LastCompletedClusterTransactionIndexWaiter.Task < index) ;
+                    
+                    ServerStore.CommandAlreadyInLog(clusterTransactionCommand, out index, out result, out _);
+                }
+                var count = ValidateClusterTransactionResult(result);
 
-            var array = new DynamicJsonArray();
-            if (clusterTransactionCommand.DatabaseCommandsCount > 0)
-            {
-                ClusterTransactionCompletionResult reply;
-                using (var timeout = new CancellationTokenSource(ServerStore.Engine.OperationTimeout))
-                using (var cts = CancellationTokenSource.CreateLinkedTokenSource(timeout.Token, HttpContext.RequestAborted))
+                if (count.HasValue) // cluster transaction contains 'put/'delete'
                 {
-                    reply = (ClusterTransactionCompletionResult)await Database.ClusterTransactionWaiter.WaitForResults(options.TaskId, cts.Token);
+                    var lastModified = Database.Time.GetUtcNow();
+                    GenerateDatabaseCommandsEvaluatedResults(clusterTransactionCommand.DatabaseCommands, index, count.Value, lastModified, options.DisableAtomicDocumentWrites, array);
                 }
-                if (reply.IndexTask != null)
+            }
+            else
+            {
+                var resultFromLeader = await ServerStore.SendToLeaderAsync(clusterTransactionCommand);
+                index = resultFromLeader.Index;
+                ValidateClusterTransactionResult(resultFromLeader.Result);
+
+                // wait for the command to be applied on this node
```

Comment:
why do we need to wait here? we have the results, no?
the only reason that I can think of, is we need to wait for indexing... 
@haludi how did we solved that in 6.0?

Reply:
We didn't 
For sharding atleast

Reply:
We don't have the results,
we can have null if the command isn't executed yet in the ClusterStateMachine